src/generate_features.py

```diff
@@ -0,0 +1,224 @@
+    df.loc[df["bathrooms"] >= 3, "bathrooms_cat"] = 3
+
+    # Create new variable to categorize beds
+    df.loc[:, "beds_cat"] = df["beds"]
+
+    # For listings where # beds = 0, assume that there is 1 bed per bedroom.
+    # If # bedrooms = 0, set as 1
+    df.loc[(df["beds"] == 0) | (df["beds"].isna()), "beds_cat"] = np.minimum(
+        np.maximum(df["bedrooms"], 1), 5
+    )
+    df.loc[df["beds"] >= 5, "beds_cat"] = 5
+
+    # Create new variable to categorize bed_type
+    df.loc[:, "bed_type_cat"] = df["bed_type"]
+    df.loc[~df["bed_type"].isin(["Real Bed"]), "bed_type_cat"] = "Other"
+
+    # Create new variable to categorize guests_included
+    df.loc[:, "guests_included_cat"] = df["guests_included"]
+    df.loc[df["guests_included"] > 2, "guests_included_cat"] = 3
+
+    # Create new variable to categorize extra_people
+    df.loc[:, "extra_people_cat"] = df["extra_people"]
+    df.loc[df["extra_people"] > 0, "extra_people_cat"] = 1
+
+    # Create variable as count of # of amenities
+    df.loc[:, "amenities_count"] = df["amenities"].str[1:-1].str.split(",").str.len()
+
+    # Convert require_guest_profile_picture to 0/1
+    df.loc[
+        df["require_guest_profile_picture"] == "t", "require_guest_profile_picture"
+    ] = 1
+    df.loc[
+        df["require_guest_profile_picture"] == "f", "require_guest_profile_picture"
+    ] = 0
+
+    # Convert require_guest_phone_verification to 0/1
+    df.loc[
+        df["require_guest_phone_verification"] == "t",
+        "require_guest_phone_verification",
+    ] = 1
+    df.loc[
+        df["require_guest_phone_verification"] == "f",
+        "require_guest_phone_verification",
+    ] = 0
+
+    return df
+
+
+def create_booking_features(df):
+    """Perform transformations on listing data to create booking features"""
+
+    # Convert instant_bookable to 0/1
+    df.loc[df["instant_bookable"] == "t", "instant_bookable"] = 1
+    df.loc[df["instant_bookable"] == "f", "instant_bookable"] = 0
+
+    # Lump super_strict_30 and super_strict_60 with strict
+    df.loc[
+        (df["cancellation_policy"] == "super_strict_30")
+        | (df["cancellation_policy"] == "super_strict_60"),
+        "cancellation_policy",
+    ] = "strict"
+
+    # Create new variable to categorize minimum_nights
+    df.loc[:, "min_nights_cat"] = df["minimum_nights"]
+    df.loc[df["minimum_nights"] < 7, "min_nights_cat"] = 1
+    df.loc[
+        (df["minimum_nights"] >= 7) & (df["minimum_nights"] < 30), "min_nights_cat",
+    ] = 2
+    df.loc[df["minimum_nights"] >= 30, "min_nights_cat"] = 3
+
+    # Create new variable to categorize maximum_nights
+    df.loc[:, "max_nights_cat"] = df["maximum_nights"]
+    df.loc[df["maximum_nights"] <= 30, "max_nights_cat"] = 1
+    df.loc[
+        (df["maximum_nights"] > 30) & (df["maximum_nights"] < 365), "max_nights_cat",
+    ] = 2
+    df.loc[df["maximum_nights"] >= 365, "max_nights_cat"] = 3
```

Comment:
Inequalities are not consistent: For minimum_nights, you use < (less than), but for maximum nights you use <= (less than or equal to). Is there a reason for this?

Reply:
This was intentional, to bucket the observations commensurate with how the values are distributed. During EDA, I looked at a histogram of the distribution of minimum nights and maximum nights, and noted where the "cutoff" points were. The partitioning can be revisited.